crates/worker/src/partition/state_machine/mod.rs

```diff
@@ -430,148 +423,177 @@ impl<Codec: RawEntryCodec> StateMachine<Codec> {
     }
 
     /// Returns the invocation in case the invocation is not a duplicate
-    async fn handle_service_invocation_idempotency_id<
-        State: IdempotencyTable + InvocationStatusTable + OutboxTable + FsmTable,
+    async fn handle_duplicated_requests<
+        State: IdempotencyTable
+            + InvocationStatusTable
+            + VirtualObjectStatusTable
+            + OutboxTable
+            + FsmTable,
     >(
         &mut self,
         ctx: &mut StateMachineApplyContext<'_, State>,
-        service_invocation: ServiceInvocation,
+        mut service_invocation: ServiceInvocation,
     ) -> Result<Option<ServiceInvocation>, Error> {
-        if let Some(idempotency_id) = service_invocation.compute_idempotency_id() {
-            if service_invocation.invocation_target.invocation_target_ty()
-                == InvocationTargetType::Workflow(WorkflowHandlerType::Workflow)
-            {
-                warn!("The idempotency key for workflow methods is ignored!");
+        let invocation_id = service_invocation.invocation_id;
+        let is_workflow_run = service_invocation.invocation_target.invocation_target_ty()
+            == InvocationTargetType::Workflow(WorkflowHandlerType::Workflow);
+        let mut has_idempotency_key = service_invocation.idempotency_key.is_some();
+
+        if !is_workflow_run && !has_idempotency_key {
+            // No duplicate id should exist
+            debug_assert!(
+                ctx.get_invocation_status(&invocation_id)
+                    .await
+                    .is_ok_and(|status| status == InvocationStatus::Free),
+                "There shouldn't be a duplicate for invocation id {}",
+                invocation_id
+            );
```

Comment:
This means that we must not retry submitting non-idempotent invocations from the ingress, right? Otherwise, we might just panic here.

Reply:
Is this strictly required or could we ignore duplicate invocations?

Reply:
> This means that we must not retry submitting non-idempotent invocations from the ingress, right? Otherwise, we might just panic here.

Yes.

> Is this strictly required or could we ignore duplicate invocations?

I guess we can ignore, but i would propose to keep this for now and we can relax later.